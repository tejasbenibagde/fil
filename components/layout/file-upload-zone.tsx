"use client";

import {
  useState,
  useRef,
  type ChangeEvent,
  type DragEvent,
  type MouseEvent,
} from "react";
import { Button } from "@/components/ui/button";
import { Upload } from "lucide-react";

interface FileUploadZoneProps {
  onFilesSelected: (files: File[]) => void;
  accept?: string;
  maxFiles?: number;
  buttonText?: string;
  description?: string;
  className?: string;
  disabled?: boolean;
}

export function FileUploadZone({
  onFilesSelected,
  accept = "*/*",
  maxFiles = 1,
  buttonText = "Browse Files",
  description = "Drag and drop your file here, or click to browse",
  className = "",
  disabled = false,
}: FileUploadZoneProps) {
  const [isDragging, setIsDragging] = useState(false);
  const fileInputRef = useRef<HTMLInputElement>(null);
  const containerRef = useRef<HTMLDivElement>(null);

  const handleDragEnter = (e: DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    if (!disabled) setIsDragging(true);
  };

  const handleDragOver = (e: DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    if (!disabled && !isDragging) setIsDragging(true);
  };

  const handleDragLeave = (e: DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();

    // Only set isDragging to false if we're leaving the container (not entering a child element)
    if (!e.currentTarget.contains(e.relatedTarget as Node)) {
      setIsDragging(false);
    }
  };

  const handleDrop = (e: DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    setIsDragging(false);

    if (disabled) return;

    const files = Array.from(e.dataTransfer.files);
    processFiles(files);
  };

  const handleFileInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    if (disabled || !e.target.files?.length) return;

    const files = Array.from(e.target.files);
    processFiles(files);
  };

  const processFiles = (files: File[]) => {
    // Filter files based on accept attribute
    let filteredFiles = files;

    if (accept !== "*/*") {
      const acceptedTypes = accept.split(",").map((type) => type.trim());
      filteredFiles = files.filter((file) => {
        return acceptedTypes.some((type) => {
          // Handle mime types (e.g., "image/*")
          if (type.endsWith("/*")) {
            const category = type.split("/")[0];
            return file.type.startsWith(`${category}/`);
          }
          // Handle specific extensions (e.g., ".pdf")
          if (type.startsWith(".")) {
            return file.name.toLowerCase().endsWith(type.toLowerCase());
          }
          // Handle exact mime types
          return file.type === type;
        });
      });
    }

    // Limit number of files
    if (maxFiles > 0 && filteredFiles.length > maxFiles) {
      filteredFiles = filteredFiles.slice(0, maxFiles);
    }

    if (filteredFiles.length > 0) {
      onFilesSelected(filteredFiles);
    }
  };

  const handleContainerClick = (e: MouseEvent<HTMLDivElement>) => {
    // Only trigger if the click is directly on the container or specific child elements
    // but not on the button or its children
    if (disabled) return;

    // Check if the click target is the button or a child of the button
    const buttonElement = containerRef.current?.querySelector("button");
    if (
      buttonElement &&
      (buttonElement === e.target || buttonElement.contains(e.target as Node))
    ) {
      return; // Don't do anything if clicking on the button
    }

    // Otherwise, trigger the file input
    if (fileInputRef.current) {
      fileInputRef.current.click();
    }
  };

  const triggerFileInput = () => {
    if (!disabled && fileInputRef.current) {
      fileInputRef.current.click();
    }
  };

  return (
    <div
      ref={containerRef}
      className={`border-2 ${
        isDragging ? "border-primary" : "border-dashed"
      } rounded-lg p-8 text-center transition-colors ${className} ${
        disabled ? "opacity-60 cursor-not-allowed" : "cursor-pointer"
      }`}
      onDragEnter={handleDragEnter}
      onDragOver={handleDragOver}
      onDragLeave={handleDragLeave}
      onDrop={handleDrop}
      onClick={handleContainerClick}
      tabIndex={0}
      role="button"
      aria-disabled={disabled}
    >
      <div className="flex flex-col items-center">
        <Upload
          className={`h-10 w-10 mb-4 ${
            isDragging ? "text-primary" : "text-muted-foreground"
          }`}
        />
        <h3 className="text-lg font-medium mb-2">
          {isDragging ? "Drop your file here" : "Upload your file"}
        </h3>
        <p className="text-sm text-muted-foreground mb-4">{description}</p>

        {/* Completely separate button without label wrapping */}
        <Button
          aria-label="upload button"
          disabled={disabled}
          onClick={(e) => {
            e.stopPropagation();
            triggerFileInput();
          }}
        >
          {buttonText}
        </Button>

        {/* Hidden file input without label */}
        <input
          data-testid="file-input"
          type="file"
          className="sr-only"
          accept={accept}
          onChange={handleFileInputChange}
          ref={fileInputRef}
          multiple={maxFiles > 1}
          disabled={disabled}
        />
      </div>
    </div>
  );
}
