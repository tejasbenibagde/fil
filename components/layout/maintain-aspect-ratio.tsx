"use client"

import { Button } from "@/components/ui/button"
import { Lock, Unlock } from "lucide-react"

interface MaintainAspectRatioProps {
  maintainAspectRatio: boolean
  toggleAspectRatio: () => void
  originalWidth: number | null
  originalHeight: number | null
  resetDimensions: () => void
  disabled?: boolean
}

export function MaintainAspectRatio({
  maintainAspectRatio,
  toggleAspectRatio,
  originalWidth,
  originalHeight,
  resetDimensions,
  disabled = false,
}: MaintainAspectRatioProps) {
  return (
    <div className="flex items-center justify-between mb-2">
      <div className="flex items-center">
        <Button
          variant="ghost"
          size="icon"
          onClick={toggleAspectRatio}
          className="mr-2"
          title={maintainAspectRatio ? "Unlock aspect ratio" : "Lock aspect ratio"}
          disabled={disabled}
          aria-label="toggle aspect ratio"
        >
          {maintainAspectRatio ? <Lock className="h-4 w-4 text-primary" /> : <Unlock className="h-4 w-4" />}
        </Button>
        <span className="text-sm font-medium">{maintainAspectRatio ? "Maintain aspect ratio" : "Free resize"}</span>
      </div>
      {originalWidth && originalHeight && (
        <Button variant="outline" size="sm" onClick={resetDimensions} disabled={disabled}>
          Reset to Original
        </Button>
      )}
    </div>
  )
}
