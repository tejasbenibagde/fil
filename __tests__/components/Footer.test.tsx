import { Footer } from '@/components/layout/footer'
import { render } from '@testing-library/react'

describe('Footer', () => {
    it('should render Footer component', () => {
        const footer = render(<Footer />)
        
        expect(footer).toMatchSnapshot()
    })
})