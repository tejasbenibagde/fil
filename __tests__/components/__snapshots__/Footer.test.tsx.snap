// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer should render Footer component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <footer
        class="w-full border-t py-6 md:py-0"
      >
        <div
          class="container flex flex-col items-center justify-between gap-4 md:h-24 md:flex-row"
        >
          <p
            class="text-center text-sm leading-loose text-muted-foreground md:text-left"
          >
            © 
            2025
             Fil. All rights reserved. Open source under MIT license.
          </p>
          <div
            class="flex items-center gap-4"
          >
            <a
              class="text-sm text-muted-foreground underline-offset-4 hover:underline"
              href="/privacy"
            >
              Privacy
            </a>
            <a
              class="text-sm text-muted-foreground underline-offset-4 hover:underline"
              href="/terms"
            >
              Terms
            </a>
            <a
              class="text-sm text-muted-foreground underline-offset-4 hover:underline"
              href="/contact"
            >
              Contact
            </a>
          </div>
        </div>
      </footer>
    </div>
  </body>,
  "container": <div>
    <footer
      class="w-full border-t py-6 md:py-0"
    >
      <div
        class="container flex flex-col items-center justify-between gap-4 md:h-24 md:flex-row"
      >
        <p
          class="text-center text-sm leading-loose text-muted-foreground md:text-left"
        >
          © 
          2025
           Fil. All rights reserved. Open source under MIT license.
        </p>
        <div
          class="flex items-center gap-4"
        >
          <a
            class="text-sm text-muted-foreground underline-offset-4 hover:underline"
            href="/privacy"
          >
            Privacy
          </a>
          <a
            class="text-sm text-muted-foreground underline-offset-4 hover:underline"
            href="/terms"
          >
            Terms
          </a>
          <a
            class="text-sm text-muted-foreground underline-offset-4 hover:underline"
            href="/contact"
          >
            Contact
          </a>
        </div>
      </div>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
